torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -7,30 +7,43 @@
 namespace torch {
 namespace jit {
 
-// Get a typekind that can be used as a key to distinguish different kinds of
-// mutable types. If the type is not mutable, return nullopt.
-//
-// TODO: We use these rules to divide wildcards into distinct "buckets", where
-// every wildcard that resolves to the same kind will alias each other. We can
-// introduce more granularity here (e.g. List[int] will never alias
-// List[float]).
-c10::optional<TypeKind> AliasDb::getMutableTypeKind(const TypePtr& type) {
+// For any mutable type, map it to a type such that all other types which it can
```

Comment:
* Yea, we don't need to unwrap the `Future`, I was following existing codebase.
Both `Future[Tensor]` and `(Tensor, int, Tensor)` are sort of special cases because these are immutable containers. So it doesn't really matter if we don't consider `Optional[Future[Tensor]]` to not alias `Future[Tensor]` so long as the aliasDb correctly considers the contained elements to maybe alias. Same thing with `(Tensor, int, Tensor) aliasing `(Tensor, Tensor, int)`. I'm not sure what the cleanest solution here is.

* Per https://github.com/pytorch/pytorch/pull/32540, Capsule Type doesn't fit in the IR.

* Currently, InterfaceTypes are handled by marking MethodCall's as conservative, so we mark all inputs as written to and add them to the wildcard set, and we cannot reorder nodes around MethodCalls. I think that approach still applies in the current approach, although it could certainly be improved.

I agree that this may require work to maintain. However, I think in our current state we're doing a sort of half-type based alias analysis that poses many of the same problems of this approach (interfaces & classes are currently consider in separate buckets) without getting a lot of the benefits. Despite my recent PR - `unifyTypes` is not a particularly hot piece of code. There aren't that many features we would add in the future that don't pose complications for our existing codebase but do for this PR.